src/Workspaces/Core/Portable/Remote/RemoteHostSessionHelpers.cs

```diff
@@ -271,15 +280,28 @@ private void OnStatusChanged(object sender, bool connection)
             Shutdown(CancellationToken.None);
         }
 
-        private void Initialize_NoLock(RemoteHostClient client, RemoteHostClient.Connection connection)
+        private void Initialize(RemoteHostClient client, RemoteHostClient.Connection connection)
         {
             Contract.ThrowIfNull(client);
             Contract.ThrowIfNull(connection);
 
-            _client = client;
-            _client.StatusChanged += OnStatusChanged;
+            lock (_gate)
+            {
+                if (_client != null)
+                {
+                    Contract.ThrowIfNull(_connectionDoNotAccessDirectly);
 
-            _connection = connection;
+                    // someone else beat us and set the connection. 
+                    // let this connection closed.
+                    connection.Dispose();
+                    return;
+                }
+
+                _client = client;
+                _client.StatusChanged += OnStatusChanged;
+
+                _connectionDoNotAccessDirectly = new ReferenceCountedDisposable<RemoteHostClient.Connection>(connection);
```

Comment:
the whole point of this is not doing await call under the lock which can cause VS to hang on shutdown. so using factory to get connection doesn't make sense.